### PR TITLE
NIAD-151: Ignore inbound message if it's a duplicate

### PR DESCRIPTION
**Pull request by bartek-sarul**

### CODE REVIEW

File: src/intTest/java/uk/nhs/digital/nhsconnect/nhais/container/ActiveMqInitializer.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.boot.test.util.TestPropertyValues;
+import org.springframework.context.ApplicationContextInitializer;
+import org.springframework.context.ConfigurableApplicationContext;
+
+@Slf4j
+public class ActiveMqInitializer implements ApplicationContextInitializer<ConfigurableApplicationContext> {
+    @Override
+    public void initialize(ConfigurableApplicationContext configurableApplicationContext) {
+        LOGGER.info("Overriding Spring Properties for mongodb and activemq !!!!!!!!!");
```
Comment (grzegorz-robak) [inline src/intTest/java/uk/nhs/digital/nhsconnect/nhais/container/ActiveMqInitializer.java]: Not enough "!" ;)

Comment (grzegorz-robak) [reply src/intTest/java/uk/nhs/digital/nhsconnect/nhais/container/ActiveMqInitializer.java]: And it's overriding properties only for ActiveMQ

File: src/intTest/java/uk/nhs/digital/nhsconnect/nhais/container/MongoDbInitializer.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.boot.test.util.TestPropertyValues;
+import org.springframework.context.ApplicationContextInitializer;
+import org.springframework.context.ConfigurableApplicationContext;
+
+@Slf4j
+public class MongoDbInitializer implements ApplicationContextInitializer<ConfigurableApplicationContext> {
+    @Override
+    public void initialize(ConfigurableApplicationContext configurableApplicationContext) {
+        LOGGER.info("Overriding Spring Properties for mongodb and activemq !!!!!!!!!");
```
Comment (grzegorz-robak) [inline src/intTest/java/uk/nhs/digital/nhsconnect/nhais/container/MongoDbInitializer.java]: It's overriding properties for MongoDB only

Comment (grzegorz-robak) [reply src/intTest/java/uk/nhs/digital/nhsconnect/nhais/container/MongoDbInitializer.java]: Might be a good idea to log actual values that are overriden

File: src/main/java/uk/nhs/digital/nhsconnect/nhais/model/edifact/Interchange.java [hunk 3]
```diff
@@ -1,14 +1,33 @@
...
 import java.util.List;
+import java.util.Optional;
 
-@Getter @Setter
+@Getter
 public class Interchange {
 
     // TODO: stub for the internal representation of an edifact interchange
-    private List<Segment> segments;
+    private final List<Segment> segments;
```
Comment (grzegorz-robak) [inline src/main/java/uk/nhs/digital/nhsconnect/nhais/model/edifact/Interchange.java:13]: Making List final doesn't make it immutable - constructor should copy or wrap it

File: src/main/java/uk/nhs/digital/nhsconnect/nhais/service/RegistrationConsumerService.java [hunk 4]
```diff
@@ -20,17 +23,27 @@
...
     public void handleRegistration(MeshMessage meshMessage) {
         LOGGER.debug("Received Registration message: {}", meshMessage);
         Interchange interchange = edifactParser.parse(meshMessage.getContent());
         LOGGER.debug("Parsed registration message into interchange: {}", interchange);
+
+        var inboundState = InboundState.fromInterchange(interchange);
+        try {
+            inboundStateRepository.save(inboundState); // this can detect duplicates as there is an unique compound index
+            LOGGER.debug("Saved inbound state: {}", inboundState);
+        } catch (DuplicateKeyException ex) {
```
Comment (grzegorz-robak) [inline src/main/java/uk/nhs/digital/nhsconnect/nhais/service/RegistrationConsumerService.java]: Shouldn't we handle this at call-site? Caller might decide what to do with duplicates (and tests would be more obvious)

**Review by grzegorz-robak — CHANGES_REQUESTED** [review]

File: src/main/java/uk/nhs/digital/nhsconnect/nhais/service/RegistrationConsumerService.java (hunk 4 above)
Comment (bartek-sarul) [reply src/main/java/uk/nhs/digital/nhsconnect/nhais/service/RegistrationConsumerService.java]: The idea is that for inbound messages (not recep) we need to skip the message if it's a duplicate. We will know if it's a duplicate when saving to DB (unique constraint violation). To be able to do that, we need to translate from mesh string message to edifact, build the inbound state record and save it.

Where would you like to put it?

Comment (grzegorz-robak) [reply src/main/java/uk/nhs/digital/nhsconnect/nhais/service/RegistrationConsumerService.java]: I thought about just not catching this exception here and leave it to caller.

File: src/intTest/java/uk/nhs/digital/nhsconnect/nhais/container/ActiveMqInitializer.java (hunk 1 above)
Comment (jaroslaw-wojciechowski) [inline src/intTest/java/uk/nhs/digital/nhsconnect/nhais/container/ActiveMqInitializer.java]: log info not valid anymore. Only activemq initialized here (similar for MongoDbInitializer.java)

**Review by jaroslaw-wojciechowski — CHANGES_REQUESTED** [review]

**Review by grzegorz-robak — DISMISSED** [review]

**Review by jaroslaw-wojciechowski — APPROVED** [review]

**Pull request merged**